geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/SchemaHelpers.scala

```diff
@@ -0,0 +1,188 @@
+  }
+
+  def randPartitionPlanner: Parser[RandomPartitionPlanner] = randPartitionPattern ^^ {
+    case d => RandomPartitionPlanner(d.toInt)
+  }
+
+  def datePlanner: Parser[DatePlanner] = datePattern ^^ {
+    case fmt => DatePlanner(DateTimeFormat.forPattern(fmt))
+  }
+
+  def geohashKeyPlanner: Parser[GeoHashKeyPlanner] = geohashPattern ^^ {
+    case o ~ b => GeoHashKeyPlanner(o, b)
+  }
+
+  // a key element consists of a separator and any number of random partitions, geohashes, and dates
+  def keypart: Parser[CompositeTextFormatter] =
+    (sep ~ rep(randEncoder | geohashEncoder | dateEncoder | constantStringEncoder )) ^^
+      {
+        case sep ~ xs => CompositeTextFormatter(xs, sep)
+      }
+
+  // the column qualifier must end with an ID-encoder
+  def cqpart: Parser[CompositeTextFormatter] =
+    phrase(sep ~ rep(randEncoder | geohashEncoder | dateEncoder | constantStringEncoder) ~ idEncoder) ^^ {
+      case sep ~ xs ~ id => CompositeTextFormatter(xs :+ id, sep)
+    }
+
+  // An index key is three keyparts, one for row, colf, and colq
+  def formatter = keypart ~ PART_DELIMITER ~ keypart ~ PART_DELIMITER ~ cqpart ^^ {
+    case rowf ~ PART_DELIMITER ~ cff ~ PART_DELIMITER ~ cqf => (rowf, cff, cqf)
+  }
+
+  // builds a geohash decoder to extract the entire geohash from the parts of the index key
+  def ghDecoderParser = keypart ~ PART_DELIMITER ~ keypart ~ PART_DELIMITER ~ cqpart ^^ {
+    case rowf ~ PART_DELIMITER ~ cff ~ PART_DELIMITER ~ cqf => {
+      val (roffset, (ghoffset, rbits)) = extractGeohashEncoder(rowf.lf, 0, rowf.sep.length)
+      val (cfoffset, (ghoffset2, cfbits)) = extractGeohashEncoder(cff.lf, 0, cff.sep.length)
+      val (cqoffset, (ghoffset3, cqbits)) = extractGeohashEncoder(cqf.lf, 0, cqf.sep.length)
+      val l = List((ghoffset, RowExtractor(roffset, rbits)),
+        (ghoffset2, ColumnFamilyExtractor(cfoffset, cfbits)),
+        (ghoffset3, ColumnQualifierExtractor(cqoffset, cqbits)))
+      GeohashDecoder(l.sortBy { case (off, _) => off }.map { case (_, e) => e })
+    }
+  }
+
+  def buildGeohashDecoder(s: String): GeohashDecoder = parse(ghDecoderParser, s).get
+
+  // extracts an entire date encoder from a key part
+  @tailrec
+  final def extractDateEncoder(seq: Seq[TextFormatter], offset: Int, sepLength: Int): Option[(String, Int)] =
+    seq match {
+      case DateTextFormatter(f)::xs => Some(f,offset)
+      case x::xs => extractDateEncoder(xs, offset + x.numBits + sepLength, sepLength)
+      case Nil => None
+    }
+
+  // extracts the geohash encoder from a keypart
+  @tailrec
+  final def extractGeohashEncoder(seq: Seq[TextFormatter], offset: Int, sepLength: Int): (Int, (Int, Int)) =
+    seq match {
+      case GeoHashTextFormatter(off, bits)::xs => (offset, (off, bits))
+      case x::xs => extractGeohashEncoder(xs, offset + x.numBits + sepLength, sepLength)
+      case Nil => (0,(0,0))
+    }
+
+  def extractIdEncoder(seq: Seq[TextFormatter], offset: Int, sepLength: Int): Int =
```

Comment:
maybe PartialFunction again... doesn't matter too much really... this may look nicer this way